### DISCUSSION TITLE
Concurrency

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-rs#60] [General]
Hi, must I wrap container with a lock in order to use it across different threads?

I saw in the code that you use some Mutex so I was thinking that maybe is secure to use it without taking care.

**Comment by GregoryConrad** [comment]
No, `Container`s are `Send + Sync` so you don't have to worry about it. The current implementation uses a `RwLock`

**Reply by busslina** [reply]
Thanks!

**Comment by busslina** [comment]
I defined the container statically using `lazy_static`. So one question arises here. The `lazy_static` [docs](https://docs.rs/lazy_static/latest/lazy_static/#semantics) says: `If the type has a destructor, then it will not run when the process exits.`

Does it affects to the container when closing the app (seems it hasn't a dispose function)? Do you have a better solution in order to have container available on the whole app?

```
lazy_static! {
    static ref CAPSULE_CONTAINER: Container = Container::new();
}
```

**Reply by GregoryConrad** [reply]
As an FYI, the community has generally accepted `once_cell` (of which a large portion has been integrated into the standard library) over `lazy_static`. In this case, you could use `LazyCell` (or maybe `LazyLock`--I'd have to check).

> Does it affects to the container when closing the app (seems it hasn't a dispose function)? Do you have a better solution in order to have container available on the whole app?

Anything defined as a `static` will leak AFAIK (without `Drop` implementation being called). You'd have to make the container elsewhere (such as beginning of `main`) in order for `Drop` to be called. But know that `Drop` is not guaranteed in Rust anyways (leaking memory is safe), so I wouldn't put too much effort into it.

**Comment by busslina** [comment]
It works fine for me using ReArch as value cacher with `Cloned` and `LazyCloned`. I just see examples like fibonnaci and it is wonderful. I don't see any effect for the Dart counterpart `use.effect`. Is not needed in Rust? Is missing?

**Reply by GregoryConrad** [reply]
You use `register(foobar())` instead of `use.foobar()` due to language differences

**Reply by busslina** [reply]
I mean the `effect` side effect. (https://github.com/GregoryConrad/rearch-dart/blob/ff73841fc354d957b1330838797723bdc857a62f/packages/rearch/lib/src/side_effects.dart#L165)

**Reply by GregoryConrad** [reply]
Side effects that have a dependencies parameter (the list of objects) doesn't map well to Rust. There is a close alternative, `run_on_change`, but I had to remove that some time back due to Rust compiler limitations:

https://github.com/GregoryConrad/rearch-rs/blob/acb177505790ca4e73c5aa1abd7a555456c44b27/rearch-effects/src/lib.rs#L142

You could always try to see if you could get that working but I'm not sure how easy it'd be. Maybe you make a `struct` for the side effect and manually `impl SideEffect` on that struct.

**Reply by busslina** [reply]
Cool, thanks! I don't needed it for now.
